### PR TITLE
Add Windows support for `//flutter/fml/backtrace.h`

### PR DESCRIPTION
**Pull request by moko256**
This PR adds Windows support for `//flutter/fml/backtrace.h`.

This PR fixes flutter/flutter#111614.

No changes in the [flutter/tests] repo.

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: fml/backtrace.cc
Comment (moko256) [inline fml/backtrace.cc:29]: Acoording to [docs](https://github.com/abseil/abseil-cpp/blob/ab2e2c4f6062999afaf960759dfccb77f350c702/absl/debugging/symbolize.h#L78), the `absl::Symbolize` is already demangling. Can we remove this function for all platforms?

File: fml/backtrace.cc
Comment (loic-sharma) [inline fml/backtrace.cc:98]:
```suggestion
```

File: fml/backtrace.cc
Comment (loic-sharma) [inline fml/backtrace.cc:23]: Please move these to line 6 above the C++ standard library headers. See: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: fml/backtrace.cc
Comment (loic-sharma) [reply fml/backtrace.cc:29]: Looking at blame history, it looks like `DemangleSymbolName` was needed by [the original implementation](https://github.com/moko256/flutter_engine/commit/7c2d975b5556c0739f98448c994f05d53e6e9c41) but should've been removed when [we moved to abseil's implementation](https://github.com/moko256/flutter_engine/commit/f4bd18b9f8a3620d64808934032b26079cd46969). @chinmaygarde @jason-simmons Is that correct?

FWIW, `absl::Symbolize` definitely demangles on Windows (referred to as [undecorating by MSVC](https://docs.microsoft.com/en-us/cpp/build/reference/decorated-names?view=msvc-170)):

1. In [initialization it sets](https://github.com/abseil/abseil-cpp/blob/ab2e2c4f6062999afaf960759dfccb77f350c702/absl/debugging/symbolize_win32.inc#L47) the [`SYMOPT_UNDNAME` option](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/symbol-options#symopt_undname)
2. [`Symbolize` calls `SymFromAddr`](https://github.com/abseil/abseil-cpp/blob/ab2e2c4f6062999afaf960759dfccb77f350c702/absl/debugging/symbolize_win32.inc#L65), which does the demangling.

File: fml/backtrace_unittests.cc
Comment (loic-sharma) [inline fml/backtrace_unittests.cc:22]: Please remove these `std::cout` lines

File: fml/backtrace_unittests.cc
Comment (loic-sharma) [inline fml/backtrace_unittests.cc:21]: Should these verify that `CanGatherBacktrace` is in the string?

**Comment by loic-sharma** [comment]
@moko256 Thanks for implementing this, it looks great! Have you verified that the CI output now contains a stack traces on Window? If no, perhaps consider creating a draft PR on the engine that contains these changes but then crashes in a test.

**Comment by loic-sharma** [comment]
Here's example output from [CI backtrace](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8802573406171423793/+/u/test:_Host_Tests_for_host_debug_unopt/stdout) (see https://github.com/flutter/engine/pull/36254):

```
[64/456] FlutterPlatformNodeDelegateTest.canGetAXNode (612 ms)
...
[----------] 1 test from FlutterPlatformNodeDelegateTest
[ RUN      ] FlutterPlatformNodeDelegateTest.canGetAXNode
[ERROR:accessibility_bridge.cc(146)] Check failed: false.
[ERROR:flutter/fml/backtrace.cc(135)] Caught signal SIGABRT during program execution.
Frame 0: 0000000142DBBDB9 abort
Frame 1: 0000000142AF8819 base::KillProcess
Frame 2: 0000000142AF87EB base::LogMessage::~LogMessage
Frame 3: 000000014023935A flutter::AccessibilityBridge::OnNodeCreated
Frame 4: 0000000142A1D8D9 ui::AXTree::NotifyNodeHasBeenReparentedOrCreated
Frame 5: 0000000142A1B6CC ui::AXTree::Unserialize
Frame 6: 000000014023840D flutter::AccessibilityBridge::CommitUpdates
Frame 7: 00000001401C0669 flutter::testing::FlutterPlatformNodeDelegateTest_canGetAXNode_Test::TestBody
Frame 8: 0000000142C80080 testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>
Frame 9: 0000000142C62511 testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>
Frame 10: 0000000142C4882F testing::Test::Run
Frame 11: 0000000142C49038 testing::TestInfo::Run
Frame 12: 0000000142C496AC testing::TestSuite::Run
Frame 13: 0000000142C53985 testing::internal::UnitTestImpl::RunAllTests
Frame 14: 0000000142C853C0 testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>
Frame 15: 0000000142C65171 testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>
Frame 16: 0000000142C53514 testing::UnitTest::Run
Frame 17: 000000014022CB41 RUN_ALL_TESTS
Frame 18: 000000014022C6C8 main
Frame 19: 0000000142D48FB4 invoke_main
Frame 20: 0000000142D48E9E __scrt_common_main_seh
Frame 21: 0000000142D48D5E __scrt_common_main
Frame 22: 0000000142D49049 mainCRTStartup
Frame 23: 00007FF86F127BD4 BaseThreadInitThunk
Frame 24: 00007FF86F2ECED1 RtlUserThreadStart

[64/456] FlutterPlatformNodeDelegateTest.canGetAXNode returned/aborted with exit code 3 (612 ms)
```

File: fml/backtrace.cc
Comment (cbracken) [reply fml/backtrace.cc:23]: `#ifdef`ed #includes go at the end, so this is correct as-is.

File: fml/backtrace.cc
Comment (cbracken) [inline fml/backtrace.cc:30]: @chinmaygarde for thoughts on preprocessor conditionals vs making a posix subclass and a windows subclass? That's something we do with things like files, mutexes etc. typically by adding a platform specific file in [`fml/platform`](https://github.com/flutter/engine/tree/main/fml/platform/) that's conditionally compiled in via the BUILD file. e.g. For files we do:
* https://github.com/flutter/engine/blob/main/fml/file.h
* https://github.com/flutter/engine/blob/main/fml/platform/posix/file_posix.cc

**Review by cbracken — COMMENTED** [review]
I haven't looked but do the existing tests cover Windows?

File: fml/backtrace.cc
Comment (loic-sharma) [inline fml/backtrace.cc:75]: Should/could we use [abseil's `GetStackTrace`](https://github.com/abseil/abseil-cpp/blob/a05366d851c5cb88065272f951e03955197e7c11/absl/debugging/stacktrace.h#L133) here?

Comment (moko256) [reply fml/backtrace.cc]: It looks nice! I'm replacing with this. I'll check output on linux...

File: fml/backtrace.cc
Comment (moko256) [reply fml/backtrace.cc]: done

Comment (moko256) [reply fml/backtrace.cc]: Without this line, the formatter makes additional whitespace to comment below to align.
```c++
    SIGALRM,  // real-time timer expired
#endif        // !FML_OS_WIN
// Here ^^^
```
Should we keep this line to remove these space, or remove this line?

File: fml/backtrace.cc
Comment (moko256) [reply fml/backtrace.cc]: I also find `absl::InstallFailureSignalHandler`. The engine doesn't use `BacktraceHere` out of signal handler, so we can completely replace backtrace.cc with `absl::InstallFailureSignalHandler`.

cc: @chinmaygarde @cbracken

Comment (moko256) [reply fml/backtrace.cc]: Source code (prototype): https://github.com/moko256/flutter_engine/commit/3ca3f107b19bfd21256657fbab9ab08954c96a20
Windows:
```
*** SIGABRT received at time=1663690051 ***
    @   000000014073933C  (unknown)  absl::AbslFailureSignalHandler
    @   0000000142957225  (unknown)  raise
    @   00000001428D5DE9  (unknown)  abort
    @   00000001400E74A9  (unknown)  fml::testing::BacktracePrinter
    @   00000001400E748E  (unknown)  fml::testing::BacktraceTest_BacktraceHere_Test::TestBody
    @   00000001427BA210  (unknown)  testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>
    @   00000001427A4201  (unknown)  testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>
    @   000000014278A59F  (unknown)  testing::Test::Run
    @   000000014278ADA8  (unknown)  testing::TestInfo::Run
    @   000000014278B41C  (unknown)  testing::TestSuite::Run
    @   00000001427956E5  (unknown)  testing::internal::UnitTestImpl::RunAllTests
    @   00000001427BE6F0  (unknown)  testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>
    @   00000001427A6E91  (unknown)  testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>
    @   0000000142795274  (unknown)  testing::UnitTest::Run
    @   0000000140279BC1  (unknown)  RUN_ALL_TESTS
    @   0000000140279848  (unknown)  main
    @   00000001428636A9  (unknown)  invoke_main
    @   000000014286358E  (unknown)  __scrt_common_main_seh
    @   000000014286344E  (unknown)  __scrt_common_main
    @   000000014286373E  (unknown)  mainCRTStartup
    @   00007FFFC2AB7034  (unknown)  BaseThreadInitThunk
    @   00007FFFC3A426A1  (unknown)  RtlUserThreadStart
```

Linux:
```
*** SIGABRT received at time=1663690758 on cpu 0 ***
PC: @     0x7f11f6efb7bb  (unknown)  raise
    @     0x55b6704b8eec         64  (unknown)
    @     0x55b6704b8c02         64  (unknown)
    @     0x7f11f7219730  1921871696  (unknown)
    @     0x55b66ff31941         32  (unknown)
    @     0x55b67205407b         96  (unknown)
    @     0x55b672046c27        112  (unknown)
    @     0x55b672036073         96  testing::Test::Run()
    @     0x55b672036734        112  testing::TestInfo::Run()
    @     0x55b672036d3d        112  testing::TestSuite::Run()
    @     0x55b6720404e1        208  testing::internal::UnitTestImpl::RunAllTests()
    @     0x55b6720571fb         96  (unknown)
    @     0x55b672048617        112  (unknown)
    @     0x55b6720400df         96  testing::UnitTest::Run()
    @     0x55b6700b8961         16  (unknown)
    @     0x55b6700b8832        720  (unknown)
    @     0x7f11f6ee809b  (unknown)  __libc_start_main
    @ 0x5541d68949564100  (unknown)  (unknown)
Aborted
```

**Comment by moko256** [comment]
I am seeing the current master's behavior on host_linux_unopt, but many functions haven't symbolized.
I ran `readelf -s -W out/host_debug_unopt/fml_unittests`, but it seems that only gtest and std's symbol are contained.
```c++
auto trace = BacktraceHere(0);
std::cout << trace << std::endl;
```
```
Frame 0: 0x55655013e0ec Unknown
Frame 1: 0x55655225eebb Unknown
Frame 2: 0x556552251a67 Unknown
Frame 3: 0x556552240eb3 testing::Test::Run()
Frame 4: 0x556552241574 testing::TestInfo::Run()
Frame 5: 0x556552241b7d testing::TestSuite::Run()
Frame 6: 0x55655224b321 testing::internal::UnitTestImpl::RunAllTests()
Frame 7: 0x55655226203b Unknown
Frame 8: 0x556552253457 Unknown
Frame 9: 0x55655224af1f testing::UnitTest::Run()
Frame 10: 0x5565502c61e1 Unknown
Frame 11: 0x5565502c60b2 Unknown
Frame 12: 0x7f442f1be09b __libc_start_main
Frame 13: 0x55655013ab6a _start
```

**Comment by cbracken** [comment]
@moko256 can you try running with `out/host_debug_unopt/exe.unstripped/fml_unittests` instead? The default Linux binaries under the build directory are stripped.

**Comment by moko256** [comment]
@cbracken I find it. Thanks.

**Comment by moko256** [comment]
I replaced signal handling with `absl::InstallFailureSignalHandler` and added a death test to check it is installed. It also works on Windows.

File: fml/backtrace.h
Comment (cbracken) [inline fml/backtrace.h:14]: Why remove this function? I presume the intent was for use when debugging.

**Review by cbracken — COMMENTED** [review]
lgtm other than the removal of `BacktraceHere`, though I'll admit I hadn't actually used it in my own debugging (I usually just use lldb if I want to grab a backtrace). 

@chinmaygarde may have thoughts on that.

File: fml/backtrace.h
Comment (moko256) [reply fml/backtrace.h:14]: Since the #16450, which introduced `BacktraceHere`, this is only used in signal handler (and test for `BacktraceHere`). I thought it was made public to test.
The handler from `absl::InstallFailureSignalHandler` also shows stacktrace when receive signal, same as from fml's `InstallCrashHandler` before. The stacktrace's format is a little different from BacktraceHere, but it seems that the existing tests doesn't depend on the format.

I just intended to avoid bitrot to follow the [style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-abandonware). I know this can be used to debug. If any engine's contributors use it to debug, I'll restore it.

**Review by loic-sharma — APPROVED** [review]
LGTM but please also get approval from @cbracken or @chinmaygarde before merging as they're more familiar with the history here

File: fml/backtrace.h
Comment (cbracken) [reply fml/backtrace.h:14]: sgtm -- I don't use it to debug, so personally I'm fine if we remove it (we can always restore it in a followup patch with appropriate tweaks if desired)

**Review by cbracken — APPROVED** [review]

File: fml/backtrace_unittests.cc
Comment (chinmaygarde) [inline fml/backtrace_unittests.cc:14]: This test does nothing useful now. Let's remove this as well as `IsCrashHandlingSupported`.

File: fml/backtrace.h
Comment (dnfield) [reply fml/backtrace.h:14]: We shouldn't have code that isn't used, but I do end up using this call frequently when debugging/testing locally. Does absl have a similar function?

**Review by chinmaygarde — APPROVED** [review]

File: fml/backtrace.h
Comment (dnfield) [reply fml/backtrace.h:14]: (This is not blocking for landing, but I'd like to know for my own edification)

Comment (cbracken) [reply fml/backtrace.h:14]: If people are using it for development, and if absl doesn't have an equivalent then let's restore it (the existing test will avoid bitrot, at least to some extent).

**Comment by moko256** [comment]
Acoording to CI results, somehow on macOS ,the `absl::GetStackTrace` writes no stacktrace and
`absl::debugging_internal::StackTraceWorksForTest` returns false.
If `absl::GetStackTrace` doesn't support on macOS, the absl's signal handler doesn't support as well.

It looks like the absil's source code contains the implementation using execinfo, and the ifdefs for macOS, but I can't investigate the actual reason right away. I'll revert the PR to the first commit, or defer to other people on investigation on macOS.

File: fml/backtrace_unittests.cc
Comment (moko256) [reply fml/backtrace_unittests.cc]: done

File: fml/backtrace.h
Comment (moko256) [reply fml/backtrace.h:14]: done

File: fml/backtrace_unittests.cc
Comment (moko256) [reply fml/backtrace_unittests.cc]: On master, the frame 0 of offset 0 is actually `CanGatherBacktrace`. I fixed the PR and added the test.

Comment (moko256) [reply fml/backtrace_unittests.cc]: Some CI tests seems to run without symbols. I added skipping the test `BacktraceStartsWithCallerName` if the function name in the first line of stacktrace is "Unknown".

Comment (moko256) [reply fml/backtrace_unittests.cc]: I tried to make all CIs pass, but only on 
Linux Unopt, execinfo's `backtrace` returns the backtrace which contains itself, and it makes hard to write this test.
I'll remove this test from PR.

**Comment by zanderso** [comment]
@cbracken @loic-sharma @chinmaygarde @dnfield From PR review triage: It looks like this PR is ready for another review pass.

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**
